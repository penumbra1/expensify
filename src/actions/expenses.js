import database from "../firebase/firebase";

/*
 * LOADING EXPENSES
 */

// Receives an array of expenses with ids
export const loadExpenses = expenses => ({
  type: "LOAD_EXPENSES",
  expenses
});

export const startLoadExpenses = () => dispatch =>
  database
    .ref("expenses")
    .once("value")
    .then(snapshot => {
      const expensesList = [];
      snapshot.forEach(child => {
        expensesList.push({ id: child.key, ...child.val() });
      });
      return dispatch(loadExpenses(expensesList));
    })
    .catch(e => console.log("Failed to load expenses from database", e));

/*
 * ADDING EXPENSES
 */

export const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});

// Instead of an action object, startAddExpense returns an "action promise"
// that takes a dispatch argument and will be processed by Thunk middleware
export const startAddExpense = (expenseData = {}) => dispatch => {
  const {
    description = "",
    note = "",
    amount = 0,
    createdAt = 0
  } = expenseData;
  const expense = { description, note, amount, createdAt };
  return database
    .ref("expenses")
    .push(expense)
    .then(data => dispatch(addExpense({ id: data.key, ...expense })))
    .catch(e => console.log("Failed to add expense", e));
};

/*
 * REMOVING EXPENSES
 */

export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

export const startRemoveExpense = ({ id }) => dispatch => {
  database
    .ref(`expenses/${id}`)
    .remove()
    .then(() => dispatch(removeExpense({ id })))
    .catch(e => console.log("Failed to remove expense", e));
};

/*
 * EDITING EXPENSES
 */

export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

export const startEditExpense = ({ id, ...updates }) => dispatch =>
  database
    .ref(`expenses/${id}`)
    .update(updates)
    .then(() => dispatch(editExpense(id, updates)))
    .catch(e => console.log("Failed to update expense", e));
