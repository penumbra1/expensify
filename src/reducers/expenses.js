const expensesDefaultState = [];

export default (state = expensesDefaultState, action) => {
  switch (action.type) {
    case "LOAD_EXPENSES":
      // Rewrite any previous state
      return action.expenses;
    case "ADD_EXPENSE":
      return [...state, action.expense];
    case "REMOVE_EXPENSE":
      return action.id ? state.filter(({ id }) => id !== action.id) : state;
    case "EDIT_EXPENSE":
      return state.map(expense => {
        if (expense.id === action.id) {
          return { ...expense, ...action.updates };
        }
        return expense;
      });
    default:
      return state;
  }
};
