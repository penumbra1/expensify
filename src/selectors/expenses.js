import moment from "moment";

export default (expenses, { text, sortBy, startDate, endDate }) => {
  const sortFunction = (a, b) => {
    switch (sortBy) {
      case "amount":
        return b.amount - a.amount;
      case "date":
        return b.createdAt - a.createdAt;
      // The store can only set sortBy to date or amount
      // Any other case would return undefined, and sort(undefined) is alphabetic
      default:
        return undefined;
    }
  };

  const filterFunction = expense => {
    const createdAtMoment = moment(expense.createdAt);
    // User can clear dates, and no date means a match for all items
    const startDateMatch = startDate
      ? startDate.isSameOrBefore(createdAtMoment, "day")
      : true;
    const endDateMatch = endDate
      ? endDate.isSameOrAfter(createdAtMoment, "day")
      : true;
    const textMatch = text
      ? expense.description.toLowerCase().includes(text.toLowerCase())
      : true;
    return startDateMatch && endDateMatch && textMatch;
  };

  return expenses.filter(filterFunction).sort(sortFunction);
};
