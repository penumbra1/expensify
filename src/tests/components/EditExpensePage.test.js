import React from "react";
import { shallow } from "enzyme";
import {
  EditExpensePage,
  mapStateToProps
} from "../../components/EditExpensePage";
import { Redirect } from "react-router-dom";
import expenses from "../fixtures/expenses";

let expense, history, editExpense, removeExpense, wrapper;

beforeEach(() => {
  expense = expenses[0];
  history = { push: jest.fn() };

  // Spies
  removeExpense = jest.fn();
  editExpense = jest.fn();

  wrapper = shallow(
    <EditExpensePage
      expense={expense}
      removeExpense={removeExpense}
      editExpense={editExpense}
      history={history}
    />
  );
});

test("should render EditExpensePage correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should redirect to NotFoundPage if expense doesn't exist", () => {
  const newProps = { expense: undefined };
  wrapper.setProps(newProps);

  expect(wrapper.find("NotFoundPage")).toHaveLength(1);
});

test("should receive correct expense prop from state and pass it down", () => {
  // Component doesn't know which expense to edit yet
  // but receives the id from router
  const expenseToEdit = expenses[2];
  const mockProps = { match: { params: { id: expenseToEdit.id } } };
  const mockState = { expenses };

  const newProps = mapStateToProps(mockState, mockProps);
  wrapper.setProps(newProps);

  expect(wrapper.find("ExpenseForm").prop("expense")).toEqual(expenseToEdit);
});

test("should call editExpense and redirect to / on valid data submission", () => {
  wrapper.find("ExpenseForm").prop("onSubmit")(expense);
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(editExpense).toHaveBeenLastCalledWith(expense.id, expense);
});

test("should call removeExpense and redirect to / on remove button click", () => {
  wrapper.find("button").simulate("click");
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(removeExpense).toHaveBeenLastCalledWith(expense);
});
