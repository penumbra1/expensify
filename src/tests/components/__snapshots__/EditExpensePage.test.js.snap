// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<Fragment>
  <ExpenseForm
    expense={
      Object {
        "amount": 1,
        "createdAt": 0,
        "description": "candy",
        "id": "first",
        "note": "sweet!",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</Fragment>
`;
